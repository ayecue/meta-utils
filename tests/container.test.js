const { Container, Signature, VariationRegistry } = require('../dist');
const AnySignatures = require('./mocks/signatures/any.json');
const GeneralSignatures = require('./mocks/signatures/general.json');
const StringSignatures = require('./mocks/signatures/string.json');
const SubStringSignatures = require('./mocks/signatures/sub-string.json');
const NotSearchableSignature = require('./mocks/signatures/not-searchable.json');
const EN = require('./mocks/descriptions/en');

describe('container', () => {
  let meta = null;

  beforeEach(() => {
    VariationRegistry.add('test.variation', [
      "hello world",
      1234,
      1337
    ]);

    meta = new Container();

    meta.addTypeSignatureFromPayload(AnySignatures);
    meta.addTypeSignatureFromPayload(GeneralSignatures);
    meta.addTypeSignatureFromPayload(StringSignatures);
    meta.addTypeSignatureFromPayload(SubStringSignatures);
    meta.addTypeSignatureFromPayload(NotSearchableSignature);

    meta.excludeFromSearch('not-searchable');

    meta.addMetaFromPayload('en', EN);
  });

  test('should return signatures', () => {
    expect(meta.getTypeSignature('general').toJSON()).toEqual(Signature.parse(GeneralSignatures).toJSON());
  });

  test('should return any signature', () => {
    expect(meta.getDefinition(['general', 'string'], 'hasIndex').getArguments().map((item) => item.getLabel())).toEqual(AnySignatures.definitions.hasIndex.arguments.map((item) => item.label));
    expect(meta.getDefinition('any', 'hasIndex').getArguments().map((item) => item.getLabel())).toEqual(AnySignatures.definitions.hasIndex.arguments.map((item) => item.label));
  });

  test('should return description', () => {
    expect(meta.getDefinition('general', 'print').getDescription()).toEqual(EN.general.print.description);
  });

  test('should return example', () => {
    expect(meta.getDefinition('general', 'print').getExample()).toEqual(EN.general.print.example);
  });

  test('should return arg default', () => {
    expect(meta.getDefinition('general', 'print').getArgument('value').getDefault().value).toEqual('');
  });

  test('should return definition', () => {
    expect(meta.getDefinition(['string'], 'split').toJSON()).toMatchSnapshot();
  });

  test('should return variations', () => {
    expect(meta.getDefinition(['sub-string'], 'myProperty').getVariations()).toEqual([1, 'foo']);
  });

  test('should return returnVariations', () => {
    expect(meta.getDefinition(['sub-string'], 'remove2').getReturnVariations()).toEqual([
      "hello world",
      1234,
      1337,
      123
    ]);
  });

  test('should return property definition', () => {
    expect(meta.getDefinition('sub-string', 'myProperty').toJSON()).toMatchSnapshot();
  });

  test('should return definitions of child definition', () => {
    expect(meta.getDefinition(['sub-string'], 'split').toJSON()).toMatchSnapshot();
  });

  test('should return definitions of parent definition', () => {
    expect(meta.getDefinition(['sub-string'], 'hasIndex').toJSON()).toMatchSnapshot();
  });

  test('should return no matches', () => {
    const result = meta.searchDefinitionMatches('not-searchable', 'hello');
    expect([...result.keys()]).toEqual([]);
  });

  test('should return two matches', () => {
    const result = meta.searchDefinitionMatches(['sub-string', 'string', 'general'], 'split');
    expect([...result.keys()]).toEqual(['sub-string', 'string']);
  });

  test('should return two matches from any', () => {
    const result = meta.searchDefinitionMatches('any', 'hasIndex');
    expect([...result.keys()]).toEqual(['general', 'string']);
  });

  test('should create fork', () => {
    const result = meta.fork();

    expect(result.signatures).toEqual(meta.signatures);
    expect(result.meta).toEqual(meta.meta);
  });

  test('should override fork but keep original', () => {
    const result = meta.fork();

    result.addMeta('en', {
      'general': {
        'print': {
          description: 'test',
          example: ['test']
        }
      }
    });

    expect(result.getDefinition('general', 'print').getDescription()).toEqual('test');
    expect(meta.getDefinition('general', 'print').getDescription()).toEqual(EN.general.print.description);
  });
});