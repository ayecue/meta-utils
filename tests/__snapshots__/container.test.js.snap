// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`container should return definition 1`] = `
Object {
  "arguments": Array [
    Object {
      "default": null,
      "label": "delimiter",
      "opt": false,
      "type": "string",
    },
  ],
  "description": "Returns a \`list\` where each item is a segment of the \`string\`. The separation depends on the provided separator \`string\`. Keep in mind that for some odd reason, this method is using regular expressions for matching.",
  "example": Array [
    "myString = \\"42 as an answer is wrong\\"",
    "segments = myString.split(\\" \\")",
    "if segments[0] != \\"42\\" then",
    "   print(\\"Invalid information spotted!\\")",
    "else",
    "   print(\\"Information seems valid!\\")",
    "end if",
  ],
  "isProtected": false,
  "returns": Array [
    Object {
      "type": "list",
      "valueType": "string",
    },
  ],
  "type": "function",
}
`;

exports[`container should return definitions of child definition 1`] = `
Object {
  "arguments": Array [
    Object {
      "default": null,
      "label": "delimiter",
      "opt": false,
      "type": "string",
    },
    Object {
      "default": null,
      "label": "extraArg",
      "opt": false,
      "type": Object {
        "keyType": "string",
        "type": "map",
        "valueType": "string",
      },
    },
  ],
  "description": null,
  "example": null,
  "isProtected": false,
  "returns": Array [
    Object {
      "type": "list",
      "valueType": "string",
    },
  ],
  "type": "function",
}
`;

exports[`container should return definitions of parent definition 1`] = `
Object {
  "arguments": Array [
    Object {
      "default": null,
      "label": "index",
      "opt": false,
      "type": "number",
    },
  ],
  "description": "Returns \`true\` if the provided index is available in the \`string\`. Otherwise, this method will return \`false\`.",
  "example": Array [
    "myString = \\"42 as an answer is wrong\\"",
    "containsIndex = myString.hasIndex(1)",
    "if containsIndex then",
    "   print(\\"String contains index of 1.\\")",
    "else",
    "   print(\\"String does not contain index of 1.\\")",
    "end if",
  ],
  "isProtected": false,
  "returns": Array [
    "number",
  ],
  "type": "function",
}
`;

exports[`container should return property definition 1`] = `
Object {
  "description": "Test, test!",
  "example": Array [
    "testing",
  ],
  "isProtected": true,
  "type": Object {
    "type": "list",
    "valueType": "string",
  },
}
`;
